### ISSUE TITLE
`.direnv` folder crashes emanote

### ISSUE THREAD

**Issue by FelipeMarcelino** [srid/emanote#563]
**Describe the bug**
Trying to execute emanote on my NixOS with flake but got the following error:
```sh
emanote: /home/felipemarcelino/Zettelkasten/.direnv/flake-inputs/lrmxkfadhlcrakza60w0mc8sc327a5qg-source/pkgs/test/nixpkgs-check-by-name/tests/symlink-invalid/pkgs/by-name/fo/foo/foo.nix: getFileStatus: does not exist (No such file or directory)
```

**To Reproduce**
Go to the folder of zettelkasten (zk) and execute Emanote 

**Expected behavior**
To have a monitored site with zettelkaten notes being monitored by Emanote regarding changes on text

**Screenshots**
![image](https://github.com/user-attachments/assets/d6640cf6-4e4e-4f7f-b94b-1dcbb3cc59ca)

**Desktop (please complete the following information):**
Firefox

**Comment by srid** [comment]
Looks like direnv's cache folder is getting in the way. 

Normally hidden folders (starting with `.`) should be ignored per:

https://github.com/srid/emanote/blob/d165ad189dbf71bd66b0c7f211c45c5431499540/emanote/src/Emanote/Source/Pattern.hs#L33-L36

resulting in you seeing instead:

![image](https://github.com/user-attachments/assets/ac64b805-3401-492e-885b-098c48aa669b)

**Comment by FelipeMarcelino** [comment]
Should I delete the direnv folder and recreate, or only delete can solve my problem?

**Comment by srid** [comment]
Until this bug gets fixed in emanote, for now - `direnv deny && rm -rf ./.direnv` should avoid the problem.

**Comment by srid** [comment]
@FelipeMarcelino Could you try 55c116182b84fd59302e38bb114fc6e7555e408b to see if this bug still happens?

**Comment by srid** [comment]
I've seen this happen on my end. @FelipeMarcelino  ?

**Comment by srid** [comment]
> [@FelipeMarcelino](https://github.com/FelipeMarcelino) Could you try [55c1161](https://github.com/srid/emanote/commit/55c116182b84fd59302e38bb114fc6e7555e408b) to see if this bug still happens?

This fixed it for me. No other reports, so closing.

**Comment by onemoresuza** [comment]
@srid, I've tested using `nix run github:srid/emanote/55c116182b84fd59302e38bb114fc6e7555e408b` and the problem still persists for me.

The only way of fixing it was by removing the `./.direnv` directory as you pointed out before.

**Comment by srid** [comment]
@onemoresuza Out of curiosity, what's your direnv version? Mine's:

```
❯ : direnv --version
2.35.0
```

**Comment by onemoresuza** [comment]
The same as mine.

**Comment by onemoresuza** [comment]
@srid, It seems that the bug only occurs if `.direnv` exists when the the server is run; if it is created after the server is already running, it is handled correctly.

These are the logs before `.direnv` is created again:

```console
❯ nix run github:srid/emanote/55c116182b84fd59302e38bb114fc6e7555e408b -- run

[Info#ema] Launching Ema under: /home/onemoresuza/Documents/zk
[Info#System.UnionMount] Traversing /home/onemoresuza/Documents/zk for files matching ["**/*.md","**/*.org","**/*.yaml","**/*.tpl","**"], ignoring ["**/.*/**","**/*~","-/**","flake.nix","flake.lock"]
[Info#System.UnionMount] Traversing /nix/store/zczs90rz2vjsfz905w9600zjki7zjgfp-emanote-1.3.18.1-data/share/ghc-9.6.6/x86_64-linux-ghc-9.6.6/emanote-1.3.18.1 for files matching ["**/*.md","**/*.org","**/*.yaml","**/*.tpl","**"], ignoring ["**/.*/**","**/*~","-/**","flake.nix","flake.lock"]
[Info#System.UnionMount] Monitoring /home/onemoresuza/Documents/zk for changes
[Info#ema] ===============================================
[Info#ema] Ema live server RUNNING: http://127.0.0.1:40163 (ws)
[Info#ema] ===============================================
[Info#System.UnionMount] Monitoring /nix/store/zczs90rz2vjsfz905w9600zjki7zjgfp-emanote-1.3.18.1-data/share/ghc-9.6.6/x86_64-linux-ghc-9.6.6/emanote-1.3.18.1 for changes
```

These are the logs after it is created (`direnv reload`):

```console
[Warn#System.UnionMount] Unhandled folder event on '.direnv' on an ignored path
[Warn#System.UnionMount] Unhandled folder event on '.direnv/flake-inputs' on an ignored path
```

**Comment by onemoresuza** [comment]
The same behavior is present on `emanote-1.2.0`, *i. e.*, the one on `nixpkgs`.

**Comment by srid** [comment]
@onemoresuza I see the `Unhandled folder event`:

<img width="857" alt="Image" src="https://github.com/user-attachments/assets/d04b7b5a-7131-40b0-9b02-cebbfc494365" />

But that's expected. And Emanote continues to run fine for me, i.e., it does not **crash** - which is what this issue is about (see title). Does it crash for you?

**Comment by srid** [comment]
> The same behavior is present on `emanote-1.2.0`, _i. e._, the one on `nixpkgs`.

I'd not recommend using nixpkgs one, by the way. It is out of date. Which reminds me that I should make a release to Hackage.

**Comment by onemoresuza** [comment]
I've checked the files about which emanote was complaining and saw that they were `git-annex` symlinks. After creating a new git branch for testing and removing them, **the problem does not persist**.

Still, I find it strange that emanote checks whether they exist or not, since they are inside a directory that should be ignored (`.direnv`). Does emanote handles/checks for symlinks differently?

**Comment by srid** [comment]
When you say "the problem does not persist", do you mean it was **crashing** before?

**Comment by onemoresuza** [comment]
Yes, I was getting the following error `emanote: /home/onemoresuza/Documents/zk/.direnv/flake-inputs/wvfcjpzzg035yrm00f0zfrq1qjvskkyw-source/assets/stj-resp-1642250.pdf: getFileStatus: does not exist (No such file or directory)`.

The `assets/stj-resp-1642250.pdf` is `git-annex` symlink.

**Comment by onemoresuza** [comment]
@srid, I believe that what causes emanote to crash is the presence of symlinks pointing to non existing files on the dot directories.

Here's a minimal example which causes emanote to crash with `emanote: /tmp/d73182c4ae5babe6/.arbitraryname/link: getFileStatus: does not exist (No such file or directory)`:

```console
./
├── .arbitraryname/
│   └── link@
│       └╶╶ ../idonotexist
├── test.md:
│ ---
│ title: A test note
│ ---
│ 
│ # This is a test note
│ 
│ A test note.
└ 
```

**Comment by onemoresuza** [comment]
@FelipeMarcelino, is `/home/felipemarcelino/Zettelkasten/.direnv/flake-inputs/lrmxkfadhlcrakza60w0mc8sc327a5qg-source/pkgs/test/nixpkgs-check-by-name/tests/symlink-invalid/pkgs/by-name/fo/foo/foo.nix` by any chance a symlink that points to a non existent file?

**Comment by srid** [comment]
@onemoresuza I'm unable to reproduce this. If you are able to reproduce, please create a step-by-step instructions.

Here's what I did:

Repro:

<img width="751" alt="Image" src="https://github.com/user-attachments/assets/23467d10-0c49-49e1-afcf-a8fb7985e1d8" />

Result:

<img width="757" alt="Image" src="https://github.com/user-attachments/assets/d105586e-8eab-4293-94aa-74480bbc5fd2" />

**Comment by onemoresuza** [comment]
@srid, by repeating your steps, I'm able to reproduce the crash, as it can be seen in this asciinema recording:

[![asciicast](https://asciinema.org/a/LJyTFWm3pabKVhyfoPc5FSvCB.svg)](https://asciinema.org/a/LJyTFWm3pabKVhyfoPc5FSvCB)

**Comment by srid** [comment]
Ah, so it happens on Linux, but not macOS. I can reproduce on Linux.

**Comment by srid** [comment]
Looks like the problem is in `unionmount`: https://github.com/srid/unionmount/pull/10

**Comment by srid** [comment]
Oh I just realized this is the same problem: https://github.com/srid/unionmount/issues/9

cc @gjz010
